# Basic Regex Homework

In [1]:
# for Python 2: use print only as a function
from __future__ import print_function

## Homework 1: FAA tower closures

A list of FAA tower closures has been copied from a [PDF](http://www.faa.gov/news/media/fct_closed.pdf) into the file **`faa.txt`**, which is stored in the **`data`** directory of the course repository.

In [2]:
# read the file into a single string
with open('faa.txt') as f:
    data = f.read()

In [3]:
# check the number of characters
len(data)

5574

In [4]:
# examine the first 500 characters
print(data[0:500])

FAA Contract Tower Closure List
(149 FCTs)
3-22-2013
LOC
ID Facility Name City State
DHN DOTHAN RGNL DOTHAN AL
TCL TUSCALOOSA RGNL TUSCALOOSA AL
FYV DRAKE FIELD FAYETTEVILLE AR
TXK TEXARKANA RGNL-WEBB FIELD TEXARKANA AR
GEU GLENDALE MUNI GLENDALE AZ
GYR PHOENIX GOODYEAR GOODYEAR AZ
IFP LAUGHLIN/BULLHEAD INTL BULLHEAD CITY AZ
RYN RYAN FIELD TUCSON AZ
FUL FULLERTON MUNI FULLERTON CA
MER CASTLE ATWATER CA
OXR OXNARD OXNARD CA
RAL RIVERSIDE MUNI RIVERSIDE CA
RNM RAMONA RAMONA CA
SAC SACRAMENTO EXECU


In [5]:
# examine the last 500 characters
print(data[-500:])

 YAKIMA WA
CWA CENTRAL WISCONSIN MOSINEE WI
EAU CHIPPEWA VALLEY RGNL EAU CLAIRE WI
ENW KENOSHA RGNL KENOSHA WI
Page 3 of 4
FAA Contract Tower Closure List
(149 FCTs)
3-22-2013
LOC
ID Facility Name City State
JVL SOUTHERN WISCONSIN RGNL JANESVILLE WI
LSE LA CROSSE MUNI LA CROSSE WI
MWC LAWRENCE J TIMMERMAN MILWAUKEE WI
OSH WITTMAN RGNL OSHKOSH WI
UES WAUKESHA COUNTY WAUKESHA WI
HLG WHEELING OHIO CO WHEELING WV
LWB GREENBRIER VALLEY LEWISBURG WV
PKB MID-OHIO VALLEY RGNL PARKERSBURG WV
Page 4 of 4



Your assignment is to **create a list of tuples** containing the **tower IDs** and the **states** they are located in.

Here is the **expected output:**

> `faa = [('DHN', 'AL'), ('TCL', 'AL'), ..., ('PKB', 'WV')]`

In [6]:
import re

In [7]:
match = re.findall(r'([A-Z]{3}) [\w -]+([A-Z]{2})',data)

In [8]:
len(match)

160

In [9]:
match[0:10]

[('DHN', 'AL'),
 ('TCL', 'AL'),
 ('FYV', 'AR'),
 ('TXK', 'AR'),
 ('GEU', 'AZ'),
 ('GYR', 'AZ'),
 ('IFP', 'IN'),
 ('EAD', 'AZ'),
 ('RYN', 'AZ'),
 ('FUL', 'CA')]

In [10]:
match[-10:]

[('EAU', 'WI'),
 ('ENW', 'WI'),
 ('JVL', 'WI'),
 ('LSE', 'WI'),
 ('MWC', 'WI'),
 ('OSH', 'WI'),
 ('UES', 'WI'),
 ('HLG', 'WV'),
 ('LWB', 'WV'),
 ('PKB', 'WV')]

As a **bonus task**, use regular expressions to extract the **number of closures** listed in the second line of the file (149), and then use an **assertion** to check that the number of closures is equal to the length of the `faa` list.

In [11]:
match = re.search('\((\d+) \w+\)',data)
print(match.group(1))

149


## Homework 2: Stack Overflow reputation

I have downloaded my **Stack Overflow reputation history** into the file **`reputation.txt`**, which is stored in the **`data`** directory of the course repository. (If you are a Stack Overflow user with a reputation of 10 or more, you should be able to [download your own reputation history](http://stackoverflow.com/reputation).)

We are only interested in the lines that **begin with two dashes**, such as:

> `-- 2012-08-30 rep +5    = 6`

That line can be interpreted as follows: "On 2012-08-30, my reputation increased by 5, bringing my reputation total to 6."

In [12]:
# read the file into a single string
with open('reputation.txt') as f:
    rep = f.read()

In [13]:
print(rep[:500])

total votes: 36
 2  12201376 (5)
-- 2012-08-30 rep +5    = 6         
 2  13822612 (10)
-- 2012-12-11 rep +10   = 16        
 2  13822612 (10)
-- 2013-03-20 rep +10   = 26        
-- 2013-12-05 rep 0     = 26        
-- 2014-01-25 rep 0     = 26        
 16  7141669 (2)
-- 2014-03-19 rep +2    = 28        
 1  12202249 (2)
-- 2014-05-11 rep +2    = 30        
 16 23599806 (2)
 2  23597220 (10)
-- 2014-05-12 rep +12   = 42        
 2  13822612 (10)
-- 2014-06-12 rep +10   = 52        
 2  2359722


In [14]:
print(rep[-500:])

-09-07 rep +10   = 318       
 3   1839257 (-1)
-- 2015-10-14 rep -1    = 317       

** rep today: 0
** rep this week (2015-11-08 - 2015-11-14): 0
** rep this month (2015-11-01 - 2015-11-30): 0
** rep this quarter (2015-10-01 - 2015-12-31): -1
** rep this year (2015-01-01 - 2015-12-31): 109
** rep from bonuses: 100
** total rep 317 :)

days represented 34
rep cap was reached via rep from upvotes *only* on 0 days
earned at least 200 reputation on 0 days
earned 10 reputation from suggested edits



Your assignment is to **create a list of tuples** containing only these dated entries, including the **date**, **reputation change** (regardless of whether it is positive/negative/zero), and **running total**.

Here is the **expected output:**

> `rep = [('2012-08-30', '+5', '6'), ('2012-12-11', '+10', '16'), ...,  ('2015-10-14', '-1', '317')]`

In [15]:
import re

In [16]:
match = re.findall(r'-- ([0-9 -]+)\w+ ([0-9+-]+)+\W+([0-9]+)',rep)
match

[('2012-08-30 ', '+5', '6'),
 ('2012-12-11 ', '+10', '16'),
 ('2013-03-20 ', '+10', '26'),
 ('2013-12-05 ', '0', '26'),
 ('2014-01-25 ', '0', '26'),
 ('2014-03-19 ', '+2', '28'),
 ('2014-05-11 ', '+2', '30'),
 ('2014-05-12 ', '+12', '42'),
 ('2014-06-12 ', '+10', '52'),
 ('2014-06-26 ', '+10', '62'),
 ('2014-07-05 ', '0', '62'),
 ('2014-09-02 ', '0', '62'),
 ('2014-09-03 ', '+10', '72'),
 ('2014-10-28 ', '0', '72'),
 ('2014-11-14 ', '+10', '82'),
 ('2014-11-18 ', '+2', '84'),
 ('2014-12-08 ', '+2', '86'),
 ('2014-12-09 ', '+10', '96'),
 ('2014-12-12 ', '+2', '98'),
 ('2014-12-24 ', '+10', '108'),
 ('2015-02-03 ', '0', '108'),
 ('2015-02-20 ', '+10', '118'),
 ('2015-03-28 ', '+10', '128'),
 ('2015-04-26 ', '+10', '138'),
 ('2015-05-05 ', '+10', '148'),
 ('2015-05-26 ', '+10', '158'),
 ('2015-05-27 ', '+20', '178'),
 ('2015-06-09 ', '0', '178'),
 ('2015-07-03 ', '+10', '188'),
 ('2015-07-06 ', '0', '188'),
 ('2015-07-22 ', '+110', '298'),
 ('2015-08-21 ', '+10', '308'),
 ('2015-09-07 ', 

As a **bonus task**, convert this list of tuples into a **pandas DataFrame**. It should have appropriate column names, and the second and third columns should be of type integer (rather than string/object).

In [17]:
import pandas as pd
df = pd.DataFrame(match, columns=['date', 'rep_change', 'rep_total'])
df.rep_change.astype(int)
df.rep_total.astype(int)

0       6
1      16
2      26
3      26
4      26
5      28
6      30
7      42
8      52
9      62
10     62
11     62
12     72
13     72
14     82
15     84
16     86
17     96
18     98
19    108
20    108
21    118
22    128
23    138
24    148
25    158
26    178
27    178
28    188
29    188
30    298
31    308
32    318
33    317
Name: rep_total, dtype: int64

In [18]:
df.head()

date rep_change rep_total
0  2012-08-30          +5         6
1  2012-12-11         +10        16
2  2013-03-20         +10        26
3  2013-12-05           0        26
4  2014-01-25           0        26